src/Streamly/Internal/Unicode/Stream.hs

```diff
@@ -423,6 +427,80 @@ data CodingFailureMode
+    = UTF8CharDecodeInit
+    | UTF8CharDecoding !DecodeState !CodePoint
+
+parseCharUtf8WithD
+    :: Monad m => CodingFailureMode -> ParserD.Parser m Word8 Char
+parseCharUtf8WithD cfm =
+    let A.Array _ ptr _ = utf8d
+    in ParserD.Parser (step' ptr) initial extract
+
+    where
+
+    {-# INLINE prefix #-}
+    prefix = "Streamly.Internal.Data.Stream.writeCharUtf8With:"
+
+    handleError table e s =
+        case cfm of
+            ErrorOnCodingFailure -> error e
+            TransliterateCodingFailure ->
+                return $ case s of
+                    Just _ -> ParserD.Done 1 replacementChar
+                    Nothing -> ParserD.Continue 0 UTF8CharDecodeInit
+            DropOnCodingFailure ->
+                case s of
+                    Just x -> step' table UTF8CharDecodeInit x
+                    Nothing -> return $ ParserD.Partial 0 UTF8CharDecodeInit
+
+    {-# INLINE initial #-}
+    initial = return $ ParserD.IPartial UTF8CharDecodeInit
+
+    {-# INLINE extract #-}
+    extract _ = error $ prefix ++ "Not enough input"
+
+    {-# INLINE step' #-}
+    step' table UTF8CharDecodeInit x =
+        -- Note: It is important to use a ">" instead of a "<=" test
+        -- here for GHC to generate code layout for default branch
+        -- prediction for the common case. This is fragile and might
+        -- change with the compiler versions, we need a more reliable
+        -- "likely" primitive to control branch predication.
+        case x > 0x7f of
+            False -> return $ ParserD.Done 0 $ unsafeChr $ fromIntegral x
+            True ->
+                let (Tuple' sv cp) = decode0 table x
+                 in case sv of
+                        12 ->
+                            let msg = prefix
+                                    ++ "Invalid first UTF8 byte" ++ show x
+                             in handleError table msg Nothing
+                        0 -> error $ prefix ++ "unreachable state"
+                        _ ->
+                            return $ ParserD.Continue 0 (UTF8CharDecoding sv cp)
+
+    step' table (UTF8CharDecoding statePtr codepointPtr) x =
+        let (Tuple' sv cp) = decode1 table statePtr codepointPtr x
+         in case sv of
+            0 -> return $ ParserD.Done 0 $ unsafeChr cp
+            12 ->
+                let msg = prefix
+                        ++ "Invalid subsequent UTF8 byte"
+                        ++ show x
+                        ++ "in state"
+                        ++ show statePtr
+                        ++ "accumulated value"
+                        ++ show codepointPtr
+                 in handleError table msg (Just x)
+            _ -> return $ ParserD.Continue 0 (UTF8CharDecoding sv cp)
+
+writeCharUtf8' :: MonadThrow m => Fold m Word8 Char
+writeCharUtf8' =  ParserD.toFold $ parseCharUtf8WithD ErrorOnCodingFailure
+
+parseCharUtf8D :: MonadThrow m => ParserD.Parser m Word8 Char
+parseCharUtf8D =  parseCharUtf8WithD TransliterateCodingFailure
```

Comment:
The parser should not use `TransliterateCodingFailure` for failure handling. It should not do any error handling, it should always use `Error` constructor on failure. We can then run an alternative parser to handle error.